libpod/runtime.go

```diff
@@ -856,6 +859,22 @@ func (r *Runtime) refresh(alivePath string) error {
 		if err := ctr.refresh(); err != nil {
 			logrus.Errorf("Refreshing container %s: %v", ctr.ID(), err)
 		}
+		// This is the only place it's safe to use ctr.state.State unlocked
+		// We're holding the alive lock, guaranteed to be the only Libpod on the system right now.
+		if ctr.AutoRemove() && ctr.state.State == define.ContainerStateExited {
```

Comment:
I don't see how this can work. resetContainerState() seems to set the state to ContainerStateConfigured if the state was not ContainerStateExited which may be fine but if you consider this is supposed to fix the case of a hard shutdown I would think the previous state could have been running, stopped or stopping.

The only case were the state is ContainerStateExited is likely when podman container cleanup would have been run successfully at which would point it should be safe to assume the the container was already deleted.

I really think the state check must happen before resetContainerState() and it should delete for all states except ContainerStateConfigured.

Reply:
The refresh logic should be forcing anything except ContainerStateConfigured and ContainerStateInitialized to ContainerStateExited. If we're not forcing Running and Stopped to Exited, that's a problem.

We should probably also catch ContainerStateRemoving now that I think about it. Pass it through unchanged and remove immediately if true.

Reply:
Yeah, our refresh logic is definitely wrong, we should be catching more than just Exited.

Reply:
Updated. We now catch anything that was running and convert it into an Exited container.